sysroot/usr/include/kernel/sys/tasks.hpp

```diff
@@ -37,26 +39,104 @@ struct px_task
     px_task_state state;
     uint64_t time_used;
     uint64_t wakeup_time;
+    const char *name;
+    px_task_alloc alloc;
 };
 
+extern px_task_t *px_current_task;
+
+#define TASK_ONLY if (px_current_task != NULL)
+
 typedef struct px_tasklist
 {
     px_task_t *head;
     px_task_t *tail;
 } px_tasklist_t;
 
+#define MAX_TASKS_QUEUED 8
+typedef struct px_tasks_sync
+{
+    px_task_t* possessor;
+    px_tasklist_t waiting;
+} px_tasks_sync_t;
 
-extern px_task_t *px_current_task;
+static inline void px_tasks_sync_init(px_tasks_sync_t *ts) {
+    *ts = {
+        .possessor = NULL,
+        .waiting = { },
+    };
+}
 
+/**
+ * @brief Initializes the kernel task manager.
+ *
+ */
 void px_tasks_init();
+/**
+ * @brief Switches to a provided task.
+ *
+ * @param task Pointer to the task struct
+ */
 extern "C" void px_tasks_switch_to(px_task_t *task);
-px_task_t *px_tasks_new(void (*entry)(void), px_task_t *storage, px_task_state state);
+/**
+ * @brief Creates a new kernel task with a provided entry point, register storage struct,
+ * and task state struct. If the storage parameter is provided, the px_task_t struct
+ * provided will be written to. If NULL is passed as the storage parameter, a pointer
+ * to a allocated task will be returned.
+ *
+ * @param entry Task function entry point
+ * @param storage Task stack structure (if NULL, a pointer to the task is returned)
+ * @param state Task state structure
+ * @return px_task_t* Pointer to the created kernel task
+ */
+px_task_t *px_tasks_new(void (*entry)(void), px_task_t *storage, px_task_state state, const char *name);
```

Comment:
Why do we need this function when we could make them constructors on the `px_task_t` type? By using an external function to "initialize" the task, that means that we have to account for potential misuse of task objects before proper initialization.

Reply:
I agree with this. See my other comment on using C++ features.

Reply:
All future discussion on this issue should be held in #176. Once a conclusion has been determined a separate MR can be opened to address the issue.

Reply:
And just for reference -- I think the struct constructors are a good idea. One less "moving part" that needs to be initialized.